File: apps/api/src/main/java/br/unifor/costify/domain/entity/Ingredient.java
package br.unifor.costify.domain.entity;

import br.unifor.costify.domain.contracts.IdGenerator;
import br.unifor.costify.domain.errors.ingredient.InvalidIngredientNameException;
import br.unifor.costify.domain.valueobject.Id;
import br.unifor.costify.domain.valueobject.Money;
import br.unifor.costify.domain.valueobject.Unit;

public class Ingredient {
  private Id id;
  private String name;
  private double packageQuantity;
  private Money packagePrice;
  private Unit packageUnit;

  /**
   * Construtor da entidade
   *
   * @param id Ingredient ID
   * @param name Ingredient name
   * @param packageQuantity total quantity in the package
   * @param packagePrice Price of the package
   * @param packageUnit Unit of measurement for the package
   */
  public Ingredient(
      Id id, String name, double packageQuantity, Money packagePrice, Unit packageUnit) {
    this.validate(name, packageQuantity);
    this.id = id;
    this.name = name;
    this.packageQuantity = packageQuantity;
    this.packagePrice = packagePrice;
    this.packageUnit = packageUnit;
  }

  /**
   * Construtor da entidade
   *
   * @param idGenerator Generator for the entity ID
   * @param name Ingredient name
   * @param packageQuantity total quantity in the package
   * @param packagePrice Price of the package
   * @param packageUnit Unit of measurement for the package
   */
  public Ingredient(
      IdGenerator idGenerator,
      String name,
      double packageQuantity,
      Money packagePrice,
      Unit packageUnit) {
    this.validate(name, packageQuantity);
    this.id = Id.generate(idGenerator);
    this.name = name;
    this.packageQuantity = packageQuantity;
    this.packagePrice = packagePrice;
    this.packageUnit = packageUnit;
  }

  public double getUnitCost() {
    return packagePrice.doubleValue() / packageUnit.toBase(packageQuantity);
  }

  // Getters
  public Id getId() {
    return id;
  }

  public String getName() {
    return name;
  }

  public double getPackageQuantity() {
    return packageQuantity;
  }

  public Money getPackagePrice() {
    return packagePrice;
  }

  public Unit getPackageUnit() {
    return packageUnit;
  }


  private void validate(String name, double packageQuantity) {
    if (name == null || name.trim().isEmpty()) {
      throw new InvalidIngredientNameException("Ingredient name cannot be null or empty");
    }
    if (packageQuantity <= 0) {
      throw new IllegalArgumentException("Package quantity must be greater than zero");
    }
  }
}

File: internal/domain/entity/ingredient.go
package entity

import (
	"strings"

	"github.com/vini/costify-go/internal/domain/contract"
	"github.com/vini/costify-go/internal/domain/errors"
	"github.com/vini/costify-go/internal/domain/valueobject"
)

// Ingredient represents an ingredient domain entity
type Ingredient struct {
	id              valueobject.Id
	name            string
	packageQuantity float64
	packagePrice    valueobject.Money
	packageUnit     valueobject.Unit
}

// NewIngredient creates a new Ingredient with explicit ID
func NewIngredient(
	id valueobject.Id,
	name string,
	packageQuantity float64,
	packagePrice valueobject.Money,
	packageUnit valueobject.Unit,
) (*Ingredient, error) {
	if err := validateIngredient(name, packageQuantity); err != nil {
		return nil, err
	}

	return &Ingredient{
		id:              id,
		name:            name,
		packageQuantity: packageQuantity,
		packagePrice:    packagePrice,
		packageUnit:     packageUnit,
	}, nil
}

// NewIngredientWithGenerator creates a new Ingredient with generated ID
func NewIngredientWithGenerator(
	idGenerator contract.IdGenerator,
	name string,
	packageQuantity float64,
	packagePrice valueobject.Money,
	packageUnit valueobject.Unit,
) (*Ingredient, error) {
	if err := validateIngredient(name, packageQuantity); err != nil {
		return nil, err
	}

	id := valueobject.Id{}.Generate(idGenerator)

	return &Ingredient{
		id:              id,
		name:            name,
		packageQuantity: packageQuantity,
		packagePrice:    packagePrice,
		packageUnit:     packageUnit,
	}, nil
}

// UnitCost calculates the cost per base unit of the ingredient
func (i *Ingredient) UnitCost() float64 {
	baseQuantity := i.packageUnit.ToBase(i.packageQuantity)
	return i.packagePrice.Amount() / baseQuantity
}

// Getters
func (i *Ingredient) Id() valueobject.Id {
	return i.id
}

func (i *Ingredient) Name() string {
	return i.name
}

func (i *Ingredient) PackageQuantity() float64 {
	return i.packageQuantity
}

func (i *Ingredient) PackagePrice() valueobject.Money {
	return i.packagePrice
}

func (i *Ingredient) PackageUnit() valueobject.Unit {
	return i.packageUnit
}

// validateIngredient validates ingredient creation parameters
func validateIngredient(name string, packageQuantity float64) error {
	if strings.TrimSpace(name) == "" {
		return errors.NewInvalidIngredientNameError("Ingredient name cannot be null or empty")
	}
	if packageQuantity <= 0 {
		return errors.NewInvalidQuantityError("Package quantity must be greater than zero")
	}
	return nil
}